index.js

```diff
@@ -1,33 +1,73 @@
 "use strict";
 
-function getNextAsyncIteratorValue(asyncIterator, index) {
-    return asyncIterator.next().then((iterator) => {
-        return { index, iterator };
-    });
+function getNextAsyncIteratorFactory(options) {
+    return async(asyncIterator, index) => {
+        try {
+            const iterator = await asyncIterator.next();
+
+            return {
+                index,
+                iterator
+            };
+        }
+        catch (err) {
+            if (options.errorCallback) {
+                options.errorCallback(err, index);
+            }
+            if (options.throwError !== false) {
+                return Promise.reject(err);
+            }
+
+            return {
+                index,
+                iterator: {
+                    done: true
+                }
+            };
+        }
+    };
 }
 
 async function* combineAsyncIterators(...iterators) {
+    let [options] = iterators;
+
+    if (typeof options.next === "function") {
+        options = {};
+    }
+    else {
+        iterators.shift();
+    }
     // Return if iterators is empty (avoid infinite loop).
     if (iterators.length === 0) {
         return;
     }
-    const promiseThatNeverResolve = new Promise(() => null);
+    const getNextAsyncIteratorValue = getNextAsyncIteratorFactory(options);
+
 
     try {
-        const asyncIteratorsValues = iterators.map(getNextAsyncIteratorValue);
+        const asyncIteratorsValues = new Map();
+        iterators.forEach((it, idx) => asyncIteratorsValues
+            .set(idx, getNextAsyncIteratorValue(it, idx))
+        );
         let numberOfIteratorsAlive = iterators.length;
```

Comment:
numberOfIteratorsAlive was useful in the old version. But now that there is a Map where we delete items you could use `asyncIteratorsValues.size` instead of numberOfIteratorsAlive (and remove it completely from the code).